### PR TITLE
Feature/proxy query

### PR DESCRIPTION
**Pull request by Dronplane**
Implementation of proxy_filter

### CODE REVIEW

File: core/search/proxy_filter.cpp [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+  lazy_filter_bitset& bitset_;
+  cost cost_;
+  document doc_;
+  doc_id_t word_idx_{0};
+  lazy_filter_bitset::word_t word_{0};
+  doc_id_t base_{doc_limits::invalid()};
+};
+
+struct proxy_query_cache {
+  iresearch_absl::flat_hash_map<const sub_reader*,
```
Comment (gnusi) [inline core/search/proxy_filter.cpp:164]: iresearch_absl -> absl

File: core/search/proxy_filter.hpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+//////////////////////////////////////////////////////////////////////////////
+class proxy_filter final : public filter {
+ public:
+  static ptr make();
+
+  using cache_ptr = std::shared_ptr<proxy_query_cache>;
+
+  static cache_ptr make_cache();
+
+  proxy_filter() noexcept : filter(irs::type<proxy_filter>::get()) {}
```
Comment (gnusi) [inline core/search/proxy_filter.hpp:51]: move to cpp?

File: core/search/proxy_filter.cpp [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  doc_iterator::ptr real_doc_itr_;
+  const document* real_doc_{nullptr};
+  size_t words_{0};
+  cost::cost_t cost_;
+};
+
+class lazy_filter_bitset_iterator final : public doc_iterator,
+                                          private util::noncopyable {
+ public:
+  lazy_filter_bitset_iterator(lazy_filter_bitset& bitset) noexcept
```
Comment (gnusi) [inline core/search/proxy_filter.cpp:93]: explicit

File: core/search/proxy_filter.cpp [hunk 4]
```diff
@@ -0,0 +1,219 @@
...
+    cost_ = cost::extract(*real_doc_itr_);
+    set_ = irs::memory::make_unique<word_t[]>(words_);
+    std::memset(set_.get(), 0, sizeof(word_t) * words_);
+    real_doc_ = irs::get<irs::document>(*real_doc_itr_);
+    begin_ = set_.get();
+    end_ = begin_;
+  }
+
+  bool get(size_t word_idx, word_t* data) {
+    constexpr auto BITS{irs::bits_required<word_t>()};
```
Comment (gnusi) [inline core/search/proxy_filter.cpp:55]: kBits

File: core/search/proxy_filter.cpp [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+    assert(set_);
+    if (word_idx >= words_) {
+      return false;
+    }
+
+    word_t* requested = set_.get() + word_idx;
+    if (requested >= end_) {
+      auto block_limit = ((word_idx + 1) * BITS) - 1;
+      auto doc_id{irs::doc_limits::invalid()};
```
Comment (gnusi) [inline core/search/proxy_filter.cpp:64]:
```suggestion
```

File: core/search/proxy_filter.cpp [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+      while (real_doc_itr_->next()) {
+        doc_id = real_doc_->value;
```
Comment (gnusi) [inline core/search/proxy_filter.cpp:66]:
```suggestion
   auto doc_id = real_doc_->value;
```

File: core/search/proxy_filter.cpp [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+  lazy_filter_bitset_iterator(lazy_filter_bitset& bitset) noexcept
+      : bitset_(bitset), cost_(bitset_.get_cost()) {
+    reset();
+  }
+
+  bool next() override {
+    while (!word_) {
+      if (bitset_.get(word_idx_, &word_)) {
+        ++word_idx_;  // move only if ok. Or we could be overflowed!
+        base_ += irs::bits_required<lazy_filter_bitset::word_t>();
```
Comment (gnusi) [inline core/search/proxy_filter.cpp:102]: `irs` namespace prefix seems to be redundant

File: tests/search/proxy_filter_test.cpp [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+INSTANTIATE_TEST_SUITE_P(
+    proxy_filter_real_filter, proxy_filter_real_filter,
+    ::testing::Combine(
+        ::testing::Values(&tests::directory<&tests::memory_directory>,
+                          &tests::directory<&tests::fs_directory>,
+                          &tests::directory<&tests::mmap_directory>),
+        ::testing::Values(tests::format_info{"1_0"},
+                          tests::format_info{"1_1", "1_0"},
+                          tests::format_info{"1_2", "1_0"},
+                          tests::format_info{"1_3", "1_0"})));
```
Comment (gnusi) [inline tests/search/proxy_filter_test.cpp:300]: 1_4/1_4simd

Comment (Dronplane) [reply tests/search/proxy_filter_test.cpp:300]: Done

**Pull request merged**